# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [1]:
import warnings
warnings.simplefilter('ignore')
import pandas as pd
import sqlite3

Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [3]:
go_sales_connect = sqlite3.connect('go_sales.sqlite')
go_staff_connect = sqlite3.connect('go_staff.sqlite')
go_crm_connect = sqlite3.connect('go_crm.sqlite')
sales = pd.read_sql_query("select * from product", go_sales_connect)
sales_country = pd.read_sql_query("select * from COUNTRY", go_sales_connect)
retailer = pd.read_sql_query("select * from retailer_site", go_crm_connect)
klant = pd.read_sql_query("select * from sales_branch", go_sales_connect)
returned_item = pd.read_sql_query("select * from returned_item", go_sales_connect)
order_details = pd.read_sql_query("select * from order_details", go_sales_connect)
staff = pd.read_sql_query("select * from sales_staff", go_staff_connect)
retailer_site = pd.read_sql_query("select * from retailer_site", go_sales_connect)
product = pd.read_sql_query("select * from product", go_sales_connect)
product_type = pd.read_sql_query("select * from product_type", go_sales_connect)


Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [9]:
pd.read_sql_query("select * from product",go_sales_connect)
pd.read_sql_query("select * from product_type",go_sales_connect)
pd.read_sql_query("select * from product_line",go_sales_connect)
pd.read_sql_query("select * from sales_staff",go_sales_connect)
pd.read_sql_query("select * from sales_branch",go_sales_connect)
pd.read_sql_query("select * from retailer_site",go_sales_connect)
pd.read_sql_query("select * from country",go_sales_connect)
pd.read_sql_query("select * from order_header",go_sales_connect)
pd.read_sql_query("select * from order_details",go_sales_connect)
pd.read_sql_query("select * from SALES_TARGETData",go_sales_connect)
pd.read_sql_query("select * from returned_item",go_sales_connect)
pd.read_sql_query("select * from return_reason",go_sales_connect)

RETURN_REASON_CODE   RETURN_DESCRIPTION_EN TRIAL888
0                  1       Defective product        T
1                  2      Incomplete product        T
2                  3   Wrong product ordered        T
3                  4   Wrong product shipped        T
4                  5  Unsatisfactory product        T

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [5]:
sales['PRODUCTION_COST'] = sales['PRODUCTION_COST'].astype(float)
kosten_overzicht = (sales['PRODUCTION_COST'] < 100) & (sales['PRODUCTION_COST'] > 50)
sales[kosten_overzicht][['PRODUCT_NAME','PRODUCTION_COST']]

PRODUCT_NAME  PRODUCTION_COST
13                Blue Steel Putter            60.00
14            Blue Steel Max Putter            81.80
17              Course Pro Golf Bag            80.00
24                  Hibernator Lite            60.00
25                       Hibernator            86.00
28  Hibernator Self - Inflating Mat            78.55
31              Hibernator Camp Cot            65.33
32     Canyon Mule Climber Backpack            62.50
56          Granite Climbing Helmet            52.86
58            Husky Harness Extreme            55.00
69                   Granite Hammer            57.13
71        TrailChef Deluxe Cook Set            85.11
81                        Polar Ice            73.33
83                     Polar Sports            80.00
84                       Polar Wave            60.00
85                    Polar Extreme            72.50
88                     Edge Extreme            80.00
91                        Seeker 35            79.19
92                        Seeker 50            92.58
93           TrailChef Double Flame            75.00
94                   Seeker Extreme            94.12
97                   Glacier Deluxe            56.25
98                      Glacier GPS            78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [6]:
sales['MARGIN'] = sales['MARGIN'].astype(float)
marge_overzicht = (sales['MARGIN'] < .2) | (sales['MARGIN'] > .6)
sales[marge_overzicht][['PRODUCT_NAME','PRODUCT_NUMBER']]



PRODUCT_NAME PRODUCT_NUMBER
29                 Hibernator Pad             21
32   Canyon Mule Climber Backpack             24
100             BugShield Natural             86
101               BugShield Spray             87
102         BugShield Lotion Lite             88
103              BugShield Lotion             89
105             BugShield Extreme             90

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [7]:
cur_is_franc = sales_country['CURRENCY_NAME'] == 'francs'
sales_country.sort_values(by="COUNTRY")[cur_is_franc]


COUNTRY_CODE      COUNTRY LANGUAGE CURRENCY_NAME TRIAL879
7            17      Belgium       EN        francs        T
0             1       France       EN        francs        T
19            8  Switzerland       EN        francs        T

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [8]:
marge_50 = sales['MARGIN'] < 0.5
unique_dates = sales[marge_50].drop_duplicates(subset='INTRODUCTION_DATE')['INTRODUCTION_DATE']
unique_dates

0      15-2-2011
3     15-12-2019
5     10-12-2019
8      5-12-2019
10    27-12-2019
15     10-1-2020
19      5-3-2013
26    26-10-2016
51     10-1-2021
99     12-6-2015
Name: INTRODUCTION_DATE, dtype: object

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [9]:
retailer_overzicht = (retailer['REGION'].notna()) & (retailer['ADDRESS2'].notna())
retailer[retailer_overzicht][['ADDRESS1','CITY']]

ADDRESS1           CITY
2            4830 Princeton    Albuquerque
5        1700 George Street  San Francisco
13        1488 Orleans Road     Charleston
28      2103 Ashbury Street    Los Angeles
29           333 Ustaff Way      San Diego
..                      ...            ...
367      10423 Kenwood Road     Cincinnati
370   332 South Main Street        Hanover
371     795 South Jefferson    Springfield
376      702 North Meridian  Oklahoma City
377  1210 Somerville Avenue    Chattanooga

[78 rows x 2 columns]

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [10]:
cur_is_dollar = (sales_country['CURRENCY_NAME'] == 'dollars') | (sales_country['CURRENCY_NAME'] == 'new dollar')
sales_country.sort_values(by="COUNTRY")[cur_is_dollar]['COUNTRY']

6         Australia
15           Canada
3            Taiwan
14    United States
Name: COUNTRY, dtype: object

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [11]:
retailer_postcode_d = (retailer['POSTAL_ZONE'].str[0] == 'D') & (retailer['ADDRESS2'].notna())
retailer[retailer_postcode_d][["ADDRESS1","ADDRESS2","CITY"]]

ADDRESS1       ADDRESS2       CITY
205  Röntgenstraße 90  3. Tür rechts  Frankfurt
351   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [12]:
returned_item['RETURN_CODE'].count()

706

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [13]:
retailer["REGION"].count()

251

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [14]:
laagste = sales['MARGIN'].min()
hoogste = sales['MARGIN'].max()
gemiddelde = (sales['MARGIN'].sum() / sales['MARGIN'].count())


Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [15]:
klant["ADDRESS2"].isna().count()

28

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [16]:
order_details['UNIT_SALE_PRICE'] = order_details['UNIT_SALE_PRICE'].astype(float)
order_details['UNIT_PRICE'] = order_details['UNIT_PRICE'].astype(float)
heeft_korting = order_details['UNIT_SALE_PRICE'] < order_details['UNIT_PRICE']
gemiddelde_korting_prijs = (order_details['UNIT_SALE_PRICE'].sum() / order_details['UNIT_SALE_PRICE'].count())
gemiddelde_korting_prijs

147.86694145786407

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [17]:
staff.groupby('POSITION_EN', as_index = False)['SALES_STAFF_CODE'].count()

POSITION_EN  SALES_STAFF_CODE
0                Branch Manager                19
1              District Manager                 4
2               General Manager                 1
3  Level 1 Sales Representative                13
4  Level 2 Sales Representative                39
5  Level 3 Sales Representative                24
6              Regional Manager                 2

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [18]:
meer_dan4 = staff.groupby('WORK_PHONE', as_index = False)['SALES_STAFF_CODE'].count()
meer_dan4 = meer_dan4['SALES_STAFF_CODE'] >= 4
staff.groupby('WORK_PHONE', as_index = False)['SALES_STAFF_CODE'].count()[meer_dan4]

WORK_PHONE  SALES_STAFF_CODE
4     +(44) 121 3505267                 5
5     +(44) 181 2033144                 4
6      +(46) 08 753 116                 4
7     +(49) 40 663 1990                 4
10  +(61) 03 2982  4242                 5
11  +31 (0)20 692 93 94                 5
12      +32 16 20.73.21                 4
13    +33 1 68 94 52 20                 5
14    +33 4 52 26 33 88                 4
18     1 (206) 292-0012                 5
19     1 (305) 557-4810                 6
20     1 (310) 281-5722                 5
21     1 (403) 232-5986                 5
22     1 (416) 493-5595                 5
23     1 (617) 268-6754                 6

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [34]:
combo_country = pd.merge(sales_country, retailer_site, left_on= 'COUNTRY_CODE', how= 'inner', right_on= 'COUNTRY_CODE')
alleen_nl = combo_country['COUNTRY'] == 'Netherlands'
combo_country[['ADDRESS1','CITY']][alleen_nl]

ADDRESS1          CITY
333                   Beets 36         Beets
334               Perestraat 1      Deventer
335  Laan van Meerdervoort 966      Den Haag
336             Kalfjeslaan 47    Amstelveen
337               Startbaan 11    Amstelveen
338      Noorderspoorsingel 28     Groningen
339              Hogehilweg 13     Amsterdam
340             Spoorstraat 43    Varsseveld
341             Westhavenweg 8     Amsterdam
342              Westzeedijk 4        Muiden
343            Gooiergracht 91         Laren
344            Hoofdstraat 150   Pieterburen
345                Olde Diek 2         Laren
346     Amsterdamsestraatweg 8      Maarssen
347        Krayenhoffstraat 11     Amsterdam
348         Overschiestraat 19     Amsterdam
349              Havenbuurt 12        Marken
350               Duivendaal 1    Wageningen
351              Voorstraat 37  Nieuweschans
352         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [33]:
combo_type = pd.merge(product, product_type, left_on= 'PRODUCT_TYPE_CODE', how= 'inner', right_on= 'PRODUCT_TYPE_CODE')
alleen_eyewear = combo_type['PRODUCT_TYPE_EN'] == 'Eyewear'
combo_type['PRODUCT_NAME'][alleen_eyewear]

87        Polar Sun
88        Polar Ice
89     Polar Sports
90       Polar Wave
91    Polar Extreme
Name: PRODUCT_NAME, dtype: object

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [40]:
combo_vestiging = pd.merge(staff, klant, left_on= 'SALES_BRANCH_CODE', how= 'inner', right_on= 'SALES_BRANCH_CODE')
BM_Vestiging = combo_vestiging['POSITION_EN'] == 'Branch Manager'
combo_vestiging[BM_Vestiging][['ADDRESS1','FIRST_NAME','LAST_NAME']]



ADDRESS1 FIRST_NAME      LAST_NAME
8                     Victoria House     Walter         Taylor
12                    3 Albany Court      Polly         Linton
20                 Interleuvenlaan 2      Gracy        Gellens
26                    Schwabentor 35      Frank          Fuchs
31                  Leopoldstraße 36      Fritz         Hirsch
34                   Piazza Duomo, 1      Maria      Iacobucci
39               Singelgravenplein 4       Kick        Kalkman
44               Isafjordsgatan 30 C      Karin      Bergström
45     75, rue du Faubourg St-Honoré      Denis           Pagé
51      7800, 756 - 6th Avenue. S.W.      Sally          White
54                  789 Yonge Street    Georges  Saint-Germain
59            1288 Dorchester Avenue       Alex      Rodriguez
65                   299 Yale Avenue   Samantha          Floyd
70                    10032 NW 186th      Julie          Olsen
75        1288 South Barrington Ave.      Corey         Wright
81               6c, rue de l'Église      Julie        Béchard
85  Prol. Paseo de la Reforma No. 51      Lucía          Reyna
92                  2315 Queen's Ave      Alice         Walter
99             Avenida Paulista, 333     Bayard          Lopes

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [1]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [2]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 